File: rust/crates/quoin-config/src/org.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Agent-IX

//! Authoring-organization resolution (quoin#381, FR-025 / FR-027).
//!
//! Precedence, first non-empty wins: an explicit `--org`, `QUOIN_ORG`, the
//! stored config, then the `origin` remote in the repository's git config.
//!
//! A stored value outranks the remote because it is something a person said,
//! where the remote is only what quoin could infer. Neither weakens the
//! no-substitution rule: when nothing yields an organization the answer is
//! [`OrgSource::None`], never a default. An org qualifies a repository so that
//! same-named repos in different organizations stay distinguishable, so
//! inventing one defeats the point of carrying it — and a declared org is
//! human-facing identity in a published artifact, where a plausible-but-wrong
//! value is harder to notice than an absent one that stops the author and asks.
//!
//! The git config is **read as a file**, never by invoking `git`, so resolution
//! holds where no git executable is present (NFR-004). This is deliberate and
//! is the reason this module does not use `gix` even though `quoin-modules`
//! does: `gix` would open a repository, and an unreadable or half-initialised
//! `.git` must be "no org here", not an error.

use std::fs;
use std::path::{Path, PathBuf};

use crate::error::ConfigError;
use crate::ids::OrgName;
use crate::paths::{Environment, RuntimeContext};
use crate::schema::QuoinConfig;
use crate::service::{ConfigService, IncidentLog};

/// Message shown when no source yielded an organization (FR-025, NFR-003).
pub const UNRESOLVED_ORG_MESSAGE: &str = concat!(
    "could not determine the authoring organization: no --org flag, no QUOIN_ORG, ",
    "no stored config, and no [remote \"origin\"] in .git/config. ",
    "Pass --org <name>, or store one with `quoin config set org <name>`."
);

/// Upper bound on a git config file read, in bytes.
///
/// `.git/config` is a small ini file. The bound stops a symlinked or corrupt
/// path being read whole; over the limit resolves to "no org here", which is
/// the same answer every other unreadable-metadata path gives.
pub const MAX_GIT_CONFIG_BYTES: u64 = 4 << 20;

/// How deep a chain of `.git` pointer files may be followed.
///
/// One hop is the real layout (worktree `.git` file → gitdir → `commondir`).
/// The ceiling exists because the pointer target is attacker-influencable
/// content on disk and a cycle would otherwise not terminate.
const MAX_GITDIR_HOPS: usize = 8;

/// Where a resolved organization came from, or `None` when nothing yielded one.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OrgSource {
    /// An explicit `--org` value.
    Flag,
    /// The `QUOIN_ORG` environment variable.
    Env,
    /// The stored configuration file.
    Config,
    /// The `origin` remote in the repository's git config.
    Git,
    /// Nothing yielded an organization.
    None,
}

impl OrgSource {
    /// The stable wire string, matching the TypeScript `OrgSource` union.
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Flag => "flag",
            Self::Env => "env",
            Self::Config => "config",
            Self::Git => "git",
            Self::None => "none",
        }
    }
}

impl std::fmt::Display for OrgSource {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_str())
    }
}

/// A resolved organization and its provenance.
///
/// The two fields are not independent: `org` is `Some` exactly when `source` is
/// not [`OrgSource::None`]. [`ResolvedOrg::org`] and [`ResolvedOrg::source`] are
/// produced together by [`resolve_org`] so the pairing cannot drift.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ResolvedOrg {
    /// The organization, absent when unresolved.
    pub org: Option<OrgName>,
    /// Where it came from.
    pub source: OrgSource,
}

impl ResolvedOrg {
    /// The unresolved answer.
    #[must_use]
    pub fn unresolved() -> Self {
        Self {
            org: None,
            source: OrgSource::None,
        }
    }

    fn found(org: OrgName, source: OrgSource) -> Self {
        Self {
            org: Some(org),
            source,
        }
    }

    /// The organization, or [`UNRESOLVED_ORG_MESSAGE`] as an error.
    ///
    /// # Errors
    /// [`ConfigError::InvalidOrgName`] when nothing resolved. Callers that must
    /// have an org use this; callers reporting provenance read the fields.
    pub fn require(&self) -> Result<&OrgName, ConfigError> {
        self.org.as_ref().ok_or(ConfigError::InvalidOrgName)
    }
}

/// Options for [`resolve_org`].
#[derive(Debug, Clone, Default)]
pub struct OrgOptions<'a> {
    /// The explicit `--org` value.
    pub flag: Option<&'a str>,
}

/// Resolve the authoring organization for the repository at `repo_root`.
///
/// `env` supplies `QUOIN_ORG` and the XDG roots; `ctx` supplies the
/// `--config-root` override and the project-local `.ix` layer. Incidents from a
/// broken config file are appended to `log` — the read still succeeds, with no
/// stored org, and resolution carries on to the remote (FR-027-AC-5).
pub fn resolve_org(
    repo_root: &Path,
    options: &OrgOptions<'_>,
    env: &dyn Environment,
    ctx: &RuntimeContext,
    log: &mut IncidentLog,
) -> ResolvedOrg {
    if let Some(flag) = options.flag.and_then(OrgName::parse_opt) {
        return ResolvedOrg::found(flag, OrgSource::Flag);
    }

    let service = ConfigService::<QuoinConfig>::for_plugin(env, ctx);
    let stored = service
        .get(log)
        .value
        .org
        .as_ref()
        .and_then(|org| OrgName::parse_opt(org.as_str()));

    if let Some(stored) = stored {
        // `QUOIN_ORG` is layered over the file by the env binding, so a value
        // here came from whichever of the two won. Report which.
        let from_env = env
            .var("QUOIN_ORG")
            .and_then(|raw| OrgName::parse_opt(&raw))
            .is_some_and(|e| e == stored);
        return ResolvedOrg::found(
            stored,
            if from_env {
                OrgSource::Env
            } else {
                OrgSource::Config
            },
        );
    }

    if let Some(from_git) = org_from_git_config(repo_root) {
        return ResolvedOrg::found(from_git, OrgSource::Git);
    }

    ResolvedOrg::unresolved()
}

/// Read the organization from the `origin` remote in `<repo_root>`'s git config.
///
/// An unreadable config, an absent `origin`, or an unparsable url yields `None`
/// — never an error, since "no org here" is a resolution outcome, not a failure.
#[must_use]
pub fn org_from_git_config(repo_root: &Path) -> Option<OrgName> {
    let git_dir = resolve_git_dir(repo_root)?;
    let path = git_dir.join("config");
    let meta = fs::symlink_metadata(&path).ok()?;
    if meta.is_file() && meta.len() > MAX_GIT_CONFIG_BYTES {
        return None;
    }
    let text = fs::read_to_string(&path).ok()?;
    origin_org(&text)
}

/// Locate the git directory holding the config for `repo_root`.
///
/// In an ordinary checkout that is `<repo_root>/.git`. In a worktree (and in a
/// submodule) `.git` is a *file* holding `gitdir: <path>`, and the config lives
/// in the shared common directory that `<gitdir>/commondir` points at — so a
/// worktree resolves the org its main checkout would.
#[must_use]
pub fn resolve_git_dir(repo_root: &Path) -> Option<PathBuf> {
    let dot_git = repo_root.join(".git");
    let meta = fs::metadata(&dot_git).ok()?;
    if meta.is_dir() {
        return Some(dot_git);
    }

    let mut base = repo_root.to_path_buf();
    let mut pointer_file = dot_git;
    for _ in 0..MAX_GITDIR_HOPS {
        let text = fs::read_to_string(&pointer_file).ok()?;
        let target = gitdir_pointer(&text)?;
        let git_dir = resolve_against(&base, Path::new(target));
        // `commondir` is how a worktree names the checkout that owns the
        // config; its absence means this gitdir holds the config itself.
        let Ok(common) = fs::read_to_string(git_dir.join("commondir")) else {
            return Some(git_dir);
        };
        let common = common.trim();
        if common.is_empty() {
            return Some(git_dir);
        }
        let resolved = resolve_against(&git_dir, Path::new(common));
        if resolved.join("config").exists() || !resolved.join(".git").is_file() {
            return Some(resolved);
        }
        base.clone_from(&resolved);
        pointer_file = resolved.join(".git");
    }
    None
}

/// Join `path` onto `base` unless it is already absolute.
fn resolve_against(base: &Path, path: &Path) -> PathBuf {
    if path.is_absolute() {
        path.to_path_buf()
    } else {
        base.join(path)
    }
}

/// Extract the target of a `gitdir: <path>` pointer line.
fn gitdir_pointer(text: &str) -> Option<&str> {
    text.lines().find_map(|line| {
        let rest = line.strip_prefix("gitdir:")?;
        let rest = rest.trim();
        (!rest.is_empty()).then_some(rest)
    })
}

/// Parse the org out of the `[remote "origin"]` url in a git config.
///
/// Handles both `git@host:org/repo.git` and `https://host/org/repo.git`. Git
/// treats section names case-insensitively but subsection names (the quoted
/// remote name) case-sensitively, and so does this.
#[must_use]
pub fn origin_org(config: &str) -> Option<OrgName> {
    let mut in_origin = false;
    for line in config.split('\n') {
        let trimmed = line.trim();
        if trimmed.starts_with('[') {
            let squashed: String = trimmed.chars().filter(|c| !c.is_whitespace()).collect();
            in_origin = squashed.to_ascii_lowercase().starts_with("[remote")
                && squashed.get("[remote".len()..) == Some("\"origin\"]");
            continue;
        }
        if !in_origin || !trimmed.starts_with("url") {
            continue;
        }
        let after_key = trimmed.get("url".len()..)?.trim_start();
        let Some(value) = after_key.strip_prefix('=') else {
            continue;
        };
        return org_from_remote_url(value.trim());
    }
    None
}

/// Extract the owning org from a remote url, or `None` when it names none.
///
/// Only *host-based* remotes carry an org: `scheme://host/org/repo` and the
/// scp-style `[user@]host:org/repo`. A local-path remote (`/srv/git/repo.git`,
/// `../sibling`, `file:///…`) has no org at all, and a host-based url with a
/// single path segment (`https://host/repo.git`) names a repo but no owner.
///
/// Both cases yield `None` rather than a best guess. Taking the second-to-last
/// path segment regardless would answer `git` for `/srv/git/repo.git`, `..` for
/// `../sibling`, and the *hostname* for `https://host/repo.git` — a
/// confidently-reported wrong org, the exact failure FR-025 exists to prevent.
#[must_use]
pub fn org_from_remote_url(url: &str) -> Option<OrgName> {
    let path = if let Some((scheme, rest)) = split_scheme(url) {
        // `file://` is a local path dressed as a url — it names no org.
        if scheme.eq_ignore_ascii_case("file") {
            return None;
        }
        let after_host = rest.find('/')?;
        &rest[after_host + 1..]
    } else if url.starts_with('/') || url.starts_with('.') || url.starts_with('~') {
        return None;
    } else {
        // scp-style `[user@]host:org/repo`. No colon at all means a relative
        // local path, which names no org.
        let colon = url.find(':')?;
        &url[colon + 1..]
    };

    let segments: Vec<&str> = path
        .trim_end_matches('/')
        .split('/')
        .filter(|s| !s.is_empty())
        .collect();
    if segments.len() < 2 {
        return None;
    }
    // The repo is the last segment; its owner is the one before it. Nested
    // namespaces (`org/subgroup/repo`) therefore qualify by the innermost
    // group, matching filament-ide-rs's repo_identity so both layers name a
    // repo alike.
    segments
        .get(segments.len() - 2)
        .and_then(|s| OrgName::parse_opt(s))
}

/// Split `scheme://rest`, if `url` has that shape.
fn split_scheme(url: &str) -> Option<(&str, &str)> {
    let idx = url.find("://")?;
    let scheme = &url[..idx];
    let mut chars = scheme.chars();
    let first = chars.next()?;
    if !first.is_ascii_alphabetic()
        || !chars.all(|c| c.is_ascii_alphanumeric() || matches!(c, '+' | '.' | '-'))
    {
        return None;
    }
    Some((scheme, &url[idx + 3..]))
}

File: src/core/org.ts
import { readFileSync, statSync } from "node:fs";
import { join, resolve } from "node:path";

import {
  ConfigService,
  getRuntimeContext,
  type RuntimeContext,
} from "@agent-ix/ix-cli-core";

import {
  QUOIN_ENV_BINDINGS,
  QUOIN_PLUGIN_ID,
  QuoinConfigSchema,
} from "../config-schema.js";
import { carriesPayload, runCoreAllowFailure } from "./exec.js";
import type {
  ResolveOrgPayload,
  ResolveOrgRequest,
  UnresolvedOrgMessagePayload,
} from "./types.js";

/**
 * Resolving the authoring organization across the `quoin-core` boundary
 * (quoin#446, Stage 7 of quoin#373).
 *
 * `src/org.ts` decided four things and this file decides none of them. The
 * precedence of flag over environment over stored config over git remote, the
 * layering and strict-schema validation of the two config documents, which
 * `[remote "origin"]` url shapes name an org, and the sentence shown when
 * nothing did — all four now live in `quoin_config` and are reached through
 * `config.resolve_org` and `config.unresolved_org_message`.
 *
 * What is left here is the host state the boundary is forbidden to acquire:
 * locating the git directory a repository's config actually lives in, and
 * reading three small files. `quoin-core`'s library half may not name a
 * filesystem (`rust/crates/quoin-core/tests/tc_library_containment.rs`), so the
 * shell does the reads and the library does the deciding. The equivalence
 * between deciding over those bytes and deciding over the tree they came from
 * is not assumed: `quoin-config`'s `tc_446_026` builds a real tree and asserts
 * both paths answer identically.
 */

/**
 * Where a resolved organization came from, or `none` when nothing yielded one.
 *
 * Reported alongside the organization itself so an author can tell an
 * explicitly stated org from one inferred off the repository's remote. The
 * five spellings are `quoin_config::OrgSource::as_str`'s; `src/write.ts` keys
 * its labels on them.
 */
export type OrgSource = "flag" | "env" | "config" | "git" | "none";

/**
 * A resolved organization: the boundary's own payload, with `source` narrowed.
 *
 * **Derived rather than re-declared, and that is FR-097 rather than taste.**
 * `src/org.ts` spelled this interface out by hand, and the fields happened to
 * match `ResolveOrgPayload` in the generated `src/core/types.ts` exactly — two
 * declarations of one contract, which is the drift `tc_1614` exists to catch
 * and did. So the shape now comes from the generated type and only the one
 * thing the JSON Schema cannot express is added: `source` is an enumeration
 * (`quoin_config::OrgSource::as_str`) that crosses the wire as a plain string,
 * and `src/write.ts` keys its labels on the five spellings.
 *
 * Consequently `degraded` is always present and `org` may be `null` rather than
 * absent — the payload's own shape. `degraded` says a config layer fell back to
 * schema defaults instead of contributing its content: `src/org.ts` discarded
 * that fact silently, so a malformed config resolved to "no stored org" with
 * nothing to show for it. The run is still a success — a broken config must not
 * stop an author writing specs (FR-027-AC-5) — but the caller can now say so.
 */
export type ResolvedOrg = Omit<ResolveOrgPayload, "source"> & {
  source: OrgSource;
};

/** Options accepted by {@link resolveOrg}. */
export interface OrgOptions {
  /** The explicit `--org` value. */
  flag?: string;
  /**
   * The environment the resolution reads. Defaults to `process.env`.
   *
   * It is a plain map layered over the config document by the declared
   * `QUOIN_ORG` binding, on the far side — one precedence rule in one place,
   * which is the property `src/org.ts` went out of its way to keep.
   */
  env?: NodeJS.ProcessEnv;
  /**
   * Project-local `.ix` layering overrides. Omit them in normal use:
   * `BaseCommand.init` publishes the runtime context (honouring
   * `--no-project-config`) and this reads it back. Supplied by tests that need
   * a hermetic root.
   */
  projectConfigRoot?: string;
  projectConfigEnabled?: boolean;
}

/**
 * Resolve the authoring organization for a repository (FR-025).
 *
 * Precedence, first non-empty wins: the explicit `--org` value, `QUOIN_ORG`,
 * the stored config, then the `origin` remote in the repository's git config.
 * Deliberately returns `source: "none"` rather than substituting a default; an
 * org qualifies a repository, so inventing one defeats the point of carrying
 * it. The rule is `quoin_config::resolve_org_from_documents`'s — this function
 * only supplies the documents it decides over.
 */
export function resolveOrg(
  repoRoot: string,
  options: OrgOptions = {},
): ResolvedOrg {
  const payload = call<ResolveOrgPayload>("config.resolve_org", {
    ...(options.flag === undefined ? {} : { flag: options.flag }),
    env: stringValuedEnv(options.env ?? process.env),
    ...documents(repoRoot, options),
  } satisfies ResolveOrgRequest);
  // Passed through rather than rebuilt: a field added to the payload reaches
  // the caller instead of being silently dropped by a literal that predates it.
  return { ...payload, source: payload.source as OrgSource };
}

/**
 * Parse the org out of the `[remote "origin"]` url in a git config.
 *
 * A `resolve_org` over that one document and nothing else: with no flag, no
 * environment and no config layer, the git remote is the only source left, so
 * this is the same rule rather than a second parser reachable another way.
 */
export function originOrg(config: string): string | undefined {
  const payload = call<ResolveOrgPayload>("config.resolve_org", {
    env: {},
    git_config: config,
  } satisfies ResolveOrgRequest);
  return payload.org ?? undefined;
}

/** Memoised: the sentence cannot change under a running binary. */
let cachedMessage: string | undefined;

/**
 * Message shown when no source yielded an organization (FR-025, NFR-003).
 *
 * Served over the boundary rather than restated here: it was a `const` in
 * `src/org.ts` and a `const` in `quoin_config::org`, and two copies of a
 * user-facing sentence is exactly the drift FR-101 retires.
 *
 * A function where `src/org.ts` exported a `const` string, and that shape
 * change is the point rather than an accident: reading the sentence is now a
 * subprocess call, and a module-level `const` would have made every importer of
 * the package spawn `quoin-core` at import time — including importers that
 * never resolve an org. The result is memoised, so the call happens at most
 * once per process and only if something asks.
 */
export function unresolvedOrgMessage(): string {
  cachedMessage ??= call<UnresolvedOrgMessagePayload>(
    "config.unresolved_org_message",
    {},
  ).message;
  return cachedMessage;
}

/** One boundary call, with the failure surfaced rather than swallowed. */
function call<T>(op: string, request: unknown): T {
  const result = runCoreAllowFailure(op, request);
  if (!carriesPayload(result.exitCode)) {
    const detail = result.diagnostics
      .map((d) => `${d.code}: ${d.message}`)
      .join("\n");
    throw new Error(
      `quoin-core ${op} exited ${result.exitCode}` +
        (detail ? `:\n${detail}` : " with no diagnostic on stderr."),
    );
  }
  return result.payload as T;
}

/**
 * `process.env` as the string→string map the request declares.
 *
 * Node types every entry as possibly `undefined` (an unset variable reads as
 * absent, not empty), and an `undefined` would fail `deny_unknown_fields`'s
 * sibling — the typed `BTreeMap<String, String>` — rather than being ignored.
 */
function stringValuedEnv(env: NodeJS.ProcessEnv): Record<string, string> {
  const out: Record<string, string> = {};
  for (const [name, value] of Object.entries(env)) {
    if (typeof value === "string") out[name] = value;
  }
  return out;
}

/** The three documents the resolution decides over, as far as they exist. */
function documents(
  repoRoot: string,
  options: OrgOptions,
): Pick<ResolveOrgRequest, "user_config" | "project_config" | "git_config"> {
  const user = readIfPresent(userConfigPath(), MAX_CONFIG_LAYER_BYTES);
  const projectPath = projectConfigPath(options);
  const project = projectPath
    ? readIfPresent(projectPath, MAX_CONFIG_LAYER_BYTES)
    : undefined;
  const gitDir = resolveGitDir(repoRoot);
  const git = gitDir
    ? readIfPresent(join(gitDir, "config"), MAX_GIT_CONFIG_BYTES)
    : undefined;
  return {
    ...(user === undefined ? {} : { user_config: user }),
    ...(project === undefined ? {} : { project_config: project }),
    ...(git === undefined ? {} : { git_config: git }),
  };
}

/**
 * The largest config layer this side will send, in bytes.
 *
 * `quoin_config::service::MAX_CONFIG_FILE_BYTES`, and the same number the
 * boundary refuses past (`ops::config::MAX_CONFIG_LAYER_BYTES`). Restated here
 * because it is a ceiling on a *read*, and `tests/core-org.test.ts` pins the
 * behaviour it produces rather than the constant.
 */
const MAX_CONFIG_LAYER_BYTES = 1 << 20;

/**
 * The largest `.git/config` this side will send, in bytes.
 *
 * `quoin_config::org::MAX_GIT_CONFIG_BYTES`, matching the in-process resolver:
 * `org_from_git_config` stats the file and answers "no org here" past this
 * number rather than reading it.
 */
const MAX_GIT_CONFIG_BYTES = 4 << 20;

/**
 * An absent, unreadable or oversized document is absent, never an error.
 *
 * "No config here" is a resolution outcome, not a failure — the same tolerance
 * `ConfigService` applies, and the reason a malformed or unreadable file
 * resolves to "no stored org" and carries on to the remote.
 *
 * **The size check is here, before the read, and that is the point.** The
 * retained in-process rule *degrades* past its ceiling — `ConfigService.get`
 * logs an incident and falls back, `org_from_git_config` returns `None`, and a
 * broken config must not stop an author writing specs (FR-027-AC-5). The
 * boundary, correctly, *refuses* an oversized field (`CORE_REFUSED`, exit 2),
 * and {@link call} throws on a refusal. Reading a 2 MiB config whole and
 * shipping it across would therefore have turned a degradation into a dead
 * `quoin write`. Over the ceiling the layer is simply absent, which is the
 * answer both in-process paths already give. The boundary's refusal stays where
 * it is, as the defence against a caller that is not this one.
 */
function readIfPresent(path: string, limit: number): string | undefined {
  try {
    // `statSync` and not `readFileSync(...).length`: a ceiling applied after
    // the whole file is in memory is a remark about an allocation that already
    // happened (rust-style §"Untrusted input").
    const stats = statSync(path);
    if (stats.isFile() && stats.size > limit) return undefined;
    return readFileSync(path, "utf8");
  } catch {
    return undefined;
  }
}

/** The user-level config file, as ix-cli-core itself spells it. */
function userConfigPath(): string {
  return ConfigService.forPlugin(QUOIN_PLUGIN_ID, QuoinConfigSchema, {
    envBindings: QUOIN_ENV_BINDINGS,
  }).filePath();
}

/**
 * The project-level config file, when a project layer applies.
 *
 * ix-cli-core declares `configPathForRoot` but does not export it from its
 * package entry, so the last segment of the layout — `<root>/config.d/<id>.yaml`
 * for every id but the reserved `core` — is spelled here. That is the one rule
 * this file restates, and `tests/core-org.test.ts` pins it by writing a file at
 * the derived path and asserting ix-cli-core's own `ConfigService` reads it, so
 * the copy cannot drift from the library that owns it without a test failing.
 */
function projectConfigPath(options: OrgOptions): string | undefined {
  const context: RuntimeContext = getRuntimeContext();
  const enabled = options.projectConfigEnabled ?? context.projectConfigEnabled;
  if (!enabled) return undefined;
  const root = options.projectConfigRoot ?? context.projectConfigRoot;
  return root === undefined
    ? undefined
    : join(root, "config.d", `${QUOIN_PLUGIN_ID}.yaml`);
}

/**
 * Locate the git directory holding the config for `repoRoot`.
 *
 * In an ordinary checkout that is `<repoRoot>/.git`. In a worktree (and in a
 * submodule) `.git` is a *file* holding `gitdir: <path>`, and the config lives
 * in the shared common directory that `<gitdir>/commondir` points at — so a
 * worktree still resolves the org its main checkout would. Kept on this side
 * because a pointer chain is host state: `quoin_config::org::resolve_git_dir`
 * is the same walk, and `tc_446_028` pins the two to the same answer.
 */
function resolveGitDir(repoRoot: string): string | undefined {
  const dotGit = join(repoRoot, ".git");
  let stats;
  try {
    stats = statSync(dotGit);
  } catch {
    return undefined;
  }
  if (stats.isDirectory()) return dotGit;

  try {
    const pointer = readFileSync(dotGit, "utf8").match(
      /^gitdir:\s*(.+)$/m,
    )?.[1];
    if (!pointer) return undefined;
    const gitDir = resolve(repoRoot, pointer.trim());
    // `commondir` is how a worktree names the checkout that owns the config;
    // its absence means this gitdir holds the config itself.
    try {
      const common = readFileSync(join(gitDir, "commondir"), "utf8").trim();
      return resolve(gitDir, common);
    } catch {
      return gitDir;
    }
  } catch {
    return undefined;
  }
}
